reana_server/scheduler.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2018 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""REANA Server Workflow Execution Scheduler."""
+
+import json
+
+from reana_commons.config import MQ_DEFAULT_QUEUES
+from reana_commons.consumer import BaseConsumer
+from reana_commons.tasks import reana_ready
+from reana_db.database import Session
+from reana_db.models import Workflow, WorkflowStatus
+
+from reana_server.api_client import current_rwc_api_client
+
+
+class WorkflowExecutionScheduler(BaseConsumer):
+    """Scheduler of workflow execution.
+
+    Class responsible for consuming from the workflow-submission queue
+    and scheduling workflows for execution based on policies and system
+    availability.
+    """
+
+    def __init__(self, **kwargs):
+        """Constructor."""
+        super(WorkflowExecutionScheduler, self).__init__(
+            queue='workflow-submission', **kwargs)
+
+    def get_consumers(self, Consumer, channel):
+        """Implement providing kombu.Consumers with queues/callbacks."""
+        return [Consumer(queues=self.queue, callbacks=[self.on_message],
+                         accept=[self.message_default_format])]
+
+    def on_message(self, workflow_submission, message):
+        """On new workflow_submission event handler."""
+        if reana_ready():
+            message.ack()
+            workflow_submission = json.loads(workflow_submission)
+            print(workflow_submission)
+            workflow_submission['status'] = 'start'
```

Comment:
I am most probably missing something here, how do we get a `status`? Seeing the call to `publish_workflow_submission` I see only `user_id`, `workflow_id_or_name`, `parameters` being passed.

Reply:
The scheduler only starts workflows, so status is always set to `start`. In brief the flow is: on_message is called when there's a new submission, if the conditions defined in the config are met then set status to `start` and call RWC, otw. requeue the message. Last part may need a bit of refactoring in another iteration because as it is mentioned in the kombu docs: http://docs.celeryproject.org/projects/kombu/en/latest/reference/kombu.message.html#kombu.message.Message.requeue it should not be used as a way to select which messages to process. Welcoming suggestions on different approaches for this task